### PR TITLE
Add C++ option report/matrix multiplication sample

### PR DESCRIPTION
**Pull request by krisdale**
# Adding a New Sample(s)

## Description

Demonstrates how compiler auto-vectorization can improve the performance of a program. The optimization report option is used to identify potential points of performance improvement.

## Checklist
Administrative
- [X] Review sample design with the appropriate: Alexandra C You

Code Development
- [ ] Implement coding guidelines and ensure code quality. [see wiki for details](https://github.com/oneapi-src/oneAPI-samples/wiki/General-Code-Guidelines)
- [X] Adhere to readme template 
- [ ] Enforce format via clang-format config file
- [X] Adhere to sample.json specification. https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-specification
- [ ] Ensure/create CI test configurations for sample (ciTests field) https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-ci-test-object
- [X] Run jsonlint on sample.json to verify json syntax. www.jsonlint.com

Security and Legal
- [ ] OSPDT Approval (see Project Manager for assistance)
- [ ] Compile using the following compiler flags and fix any warnings, the falgs are: "/Wall -Wformat-security -Werror=format-security"
- [ ] Bandit Scans (Python only)
- [ ] Virus scan

Review
- [ ] Review DPC++ code with Paul Peterseon. (GitHub User: pmpeter1)
- [ ] Review readme with Tom Lenth(@tomlenth) and/or Project Manager
- [ ] Tested using Dev Cloud when applicable

### CODE REVIEW

**Comment by jimmytwei** [comment]
@krisdale I was able to compile and run the samples. This will go out with the 2024.0 release. If you want this to go out with the 2023.2 AI Kit release, please update the PR to use the ``2023.2_AIKit`` branch.

File: DirectProgramming/C++/CompilerInfrastructure/guided_matmul_opt_report/sample.json [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  "targetDevice": ["CPU"],
+  "builder": ["make"],
+  "languages": [{"cpp":{}}],
+  "ciTests": {
+    "linux": [
+      {
+        "id": "baseline",
+        "steps": [
+          "make clean",
+          "make baseline"
```
Comment (jimmytwei) [inline DirectProgramming/C++/CompilerInfrastructure/guided_matmul_opt_report/sample.json:17]: @krisdale If you want the CI to run the sample after running make (to verify that it works), I would add ``./MatVector``

File: DirectProgramming/C++/CompilerInfrastructure/guided_matmul_opt_report/sample.json [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        ]
+      },
+      {
+        "id": "vec_report1",
+        "steps": [
+          "make clean",
+          "make vec_report1"
```
Comment (jimmytwei) [inline DirectProgramming/C++/CompilerInfrastructure/guided_matmul_opt_report/sample.json:24]: @krisdale If you want the CI to run the sample after running make (to verify that it works), I would add ``./vec_samples_report1.out``

File: DirectProgramming/C++/CompilerInfrastructure/guided_matmul_opt_report/sample.json [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        ]
+      },
+      {
+        "id": "vec_report2",
+        "steps": [
+          "make clean",
+          "make vec_report2"
```
Comment (jimmytwei) [inline DirectProgramming/C++/CompilerInfrastructure/guided_matmul_opt_report/sample.json:31]: @krisdale If you want the CI to run the sample after running make (to verify that it works), I would add ``./vec_samples_report2.out``

File: DirectProgramming/C++/CompilerInfrastructure/guided_matmul_opt_report/sample.json [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        ]
+      },
+      {
+        "id": "vec_report3",
+        "steps": [
+          "make clean",
+          "make vec_report3"
```
Comment (jimmytwei) [inline DirectProgramming/C++/CompilerInfrastructure/guided_matmul_opt_report/sample.json:38]: @krisdale If you want the CI to run the sample after running make (to verify that it works), I would add ``./vec_samples_report3.out``

**Review by jimmytwei — APPROVED** [review]

**Comment by jimmytwei** [comment]
@tomlenth Added you on to review the README. If there's someone else that should review this, please add them on here.

**Comment by krisdale** [comment]
@jimmytwei Thanks for the review - updates applied as you suggested.

**Review by jimmytwei — APPROVED** [review]

**Review by jimmytwei — APPROVED** [review]

**Pull request merged**